plugin/trino-memsql/src/main/java/io/trino/plugin/memsql/MemSqlTimeColumnMapping.java

```diff
@@ -0,0 +1,98 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.memsql;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.trino.plugin.jdbc.ColumnMapping;
+import io.trino.plugin.jdbc.LongReadFunction;
+import io.trino.plugin.jdbc.LongWriteFunction;
+import io.trino.spi.type.TimeType;
+
+import java.time.LocalTime;
+import java.time.format.DateTimeParseException;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static io.trino.plugin.jdbc.StandardColumnMappings.fromTrinoTime;
+import static io.trino.spi.type.Timestamps.NANOSECONDS_PER_DAY;
+import static io.trino.spi.type.Timestamps.PICOSECONDS_PER_DAY;
+import static io.trino.spi.type.Timestamps.PICOSECONDS_PER_NANOSECOND;
+import static io.trino.spi.type.Timestamps.round;
+import static java.time.format.DateTimeFormatter.ISO_LOCAL_TIME;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * {@link MemSqlTimeColumnMapping} is a utility class for parsing times returned by {@code SingleStore} driver.
+ *
+ * {@see com.singlestore.jdbc.codec.list.LocalTimeCodec}
+ */
+class MemSqlTimeColumnMapping
+{
+    static ColumnMapping memSqlTimeColumnMapping(TimeType timeType)
+    {
+        return ColumnMapping.longMapping(
+                timeType,
+                memsqlTimeReadFunction(timeType),
+                memsqlTimeWriteFunction(timeType.getPrecision()));
+    }
+
+    static LongReadFunction memsqlTimeReadFunction(TimeType timeType)
+    {
+        requireNonNull(timeType, "timeType is null");
+        checkArgument(timeType.getPrecision() <= 9, "Unsupported type precision: %s", timeType);
+        int precision = timeType.getPrecision();
+        return (resultSet, columnIndex) -> {
+            // SingleStore driver wraps values <00:00:00 and >24:00:00, but we must disallow them
+            // LocalTime time = resultSet.getObject(columnIndex, LocalTime.class);
+            String timeString = resultSet.getString(columnIndex);
+            return parseMemSqlLocalTime(timeString, precision);
+        };
+    }
+
+    @VisibleForTesting
+    static long parseMemSqlLocalTime(String timeString, int precision)
+    {
+        try {
+            LocalTime time = LocalTime.from(ISO_LOCAL_TIME.parse(timeString));
+            long nanosOfDay = time.toNanoOfDay();
+            verify(nanosOfDay < NANOSECONDS_PER_DAY, "Invalid value of nanosOfDay: %s", nanosOfDay);
+            long picosOfDay = nanosOfDay * PICOSECONDS_PER_NANOSECOND;
+            long rounded = round(picosOfDay, 12 - precision);
+            if (rounded == PICOSECONDS_PER_DAY) {
+                rounded = 0;
+            }
+            return rounded;
+        }
+        catch (DateTimeParseException e) {
+            throw new IllegalStateException(e.getMessage(), e);
```

Comment:
@grantatspothero (or any reviewers) - any thoughts on what Exception(s) we should be throwing here, and whether we need to wrap the original `DateTimeParseException` ?

Are there rules/guidelines for keeping our exception messages "consistent" (between connectors, or within a single connector)? Or by not catching underlying exceptions, do we not *really* control / guarantee the *exact* type of RuntimeException that will be thrown, so we don't care if they change? (e.g., here we're throwing to prevent SingleStore wrapping a value outside the allowable range - so we throw a different RuntimeException to that which MariaDb's TimeCodec would have thrown - are these not considered part of any API guarantees?)

Reply:
I do not think we consider exceptions as part as the API for most places in Trino. Usually Optional/a custom result class is used if you want to encode failure into the API.  

For this specific case of column mappings, standard column mappings can throw multiple types of errors, see: 
https://github.com/trinodb/trino/blob/master/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java#L201-L202

Which throws both a `NullPointerException` or an `IllegalArgumentException`

Reply:
I'd be curious to hear more about when to use `TrinoException` vs other types of runtime exceptions though.

Reply:
For user-facing errors it must be TrinoException since there is handling for those in the engine. For things which aren't allowed or unexpected or developer errors a proper runtime exception.

For example if we reach the default block of an exhaustive switch it's most likely developer error - the user didn't do anything wrong so an IllegalStateException makes sense.

Reply:
It'd be nice to have a "Strict Mode" switch we could turn on to identify issues around missing mappings - we could be interpreting fields incorrectly, sending back garbage, and have no idea because the garbage fits in the expected range.

Reply:
Strict mode == type mapping tests.

Unfortunately no one opts into features so we won't get any feedback at all even if a strict mode existed. You would notice that when adding new features we usually enable them by default and add a killswitch if needed (instead of disabling by default).

Thanks for raising this, this is an important part of the development philosophy and isn't documented. I'll send a PR.